### PR TITLE
Translation support for system messages in groups

### LINKED ISSUE #5757: system messages in party chat are not translated (skills castings, quest progress / damage reports)
Hello, can you somehow translate it in Russian and other languages? And other similar messages.
[/website/src/controllers/user.js#L462](https://github.com/HabitRPG/habitrpg/blob/68f5a162fd637938c4bd910daf212dba45a66ecd/website/src/controllers/user.js#L462)

```
var message = '`'+user.profile.name+' casts '+spell.text() + (targetType=='user' ? ' on '+found.profile.name : ' for the party')+'.`';
```
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/25536352-system-messages-in-party-chat-are-not-translated-skills-castings-quest-progress-damage-reports?utm_campaign=plugin&utm_content=tracker%2F68393&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F68393&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Yutsuten**
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)

[//]: # (Put Issue # or URL here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #5757

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)
- System messages in groups now are `i18n.t` strings
- The translation is done on the fly, so members on the same party using different languages will see the messages in their respective language
- Translations done from server side (on API)
- Old messages still work, but aren't translated
- Added a new object `info` inside the `chat` object in the database

#### Before change example:
```json
{
  "id" : "4a1df240-a8c6-4e23-8644-a2cb9bef9f47",
  "text" : "`user2 attacks The Basi-List for 0.5 damage.` `The Basi-List attacks party for 0.0 damage.`",
  "timestamp" : 1518349031167,
  "likes" : { },
  "flags" : { },
  "flagCount" : 0,
  "uuid" : "system"
}
```

#### After change example:
```json
{
  "id" : "4a1df240-a8c6-4e23-8644-a2cb9bef9f47",
  "text" : "`user2 attacks The Basi-List for 0.5 damage.` `The Basi-List attacks party for 0.0 damage.`",
  "info": {
    "type": "boss_damage",
    "user": "user2",
    "quest": "an_identifier",
    "userDamage": 0.5,
    "bossDamage": 0.0
  },
  "timestamp" : 1518349031167,
  "likes" : { },
  "flags" : { },
  "flagCount" : 0,
  "uuid" : "system"
}
```

[//]: # (Put User ID in here - found on the Habitica website at User Icon > Settings > API)

----
UUID: 203f208a-e0f0-4442-bf16-d4f19e553e8a

### CODE REVIEW

File: website/server/models/group.js
Comment (paglias) [inline website/server/models/group.js]: two things: can you remove Bluebird and switch to native Promises? We're going to move away from Bluebird in the coming weeks. And I'm a bit worried about performances here, can we cache the username inside the info object instead of fetching it every time the chat is loaded?

Comment (Yutsuten) [reply website/server/models/group.js]: So you think it is better to save the username instead of the uuid? I saved the uuid to update the username automatically if the user updates it, but if you think it does not worth the extra processing, I can change it to the username.
If I do this change, the Promise won't be needed.

Comment (Alys) [reply website/server/models/group.js]: @Yutsuten It's actually better to display the profile name that the player was using at the time that the boss damage was done. Chat messages made by the player themself always show the name at the time the post was made - they aren't updated later if the player changes their name.

File: website/common/locales/en/groups.json
Comment (Alys) [inline website/common/locales/en/groups.json]: I think most of these new strings would be best placed in `quests.json`. The two strings for skills could go in `character.json`, immediately under the current `youCastParty` string (around line 172).

File: website/server/models/group.js
Comment (paglias) [inline website/server/models/group.js]: can you add a comment explaining whats' stored in message.info?

**Review by paglias — CHANGES_REQUESTED** [review]

**Comment by paglias** [comment]
There's a conflict too

**Comment by Yutsuten** [comment]
Looks like the tests are failing at branch `develop`...

**Comment by paglias** [comment]
@Yutsuten no worries about that, it happens some times. The team is busy this week but I hope to be able to review this this or next week

**Comment by Yutsuten** [comment]
Got it. Thank you!

File: website/common/locales/en/character.json
Comment (khipkin) [inline website/common/locales/en/character.json]: nit/suggestion: the `youCastTarget` and `youCastParty` strings above use the `spell` variable, so also use `spell` (not `spellName`) here for consistency

File: website/common/locales/en/quests.json
Comment (khipkin) [inline website/common/locales/en/quests.json]: grammar: missing a comma before "and it doesn't want to hurt anyone unfairly."

File: website/server/controllers/api-v3/user/spells.js
Comment (khipkin) [inline website/server/controllers/api-v3/user/spells.js]: It would be awesome if all of these hard-coded messages could use the new localized strings instead, even if the `en` lang is still explicitly used, so these strings don't have to be hunted down and changed every time the localized strings get updated.

ex: website/server/conrollers/api-v3/tasks/groups.js (line 201)

**Review by khipkin — CHANGES_REQUESTED** [review]
Please take all of my comments as just suggestions until staff comments otherwise.

**Comment by Yutsuten** [comment]
@khipkin I agree with all you suggestions! Making the changes here...

**Comment by Yutsuten** [comment]
This time the tests finished with an interesting error, I've never seen it before.
Should we consider it?

```
[15:04:09] Finished 'test:api-v3:integration' after 11 min

The command "npm run $TEST" exited with 0.
$ if [ $COVERAGE ]; then ./node_modules/.bin/lcov-result-merger 'coverage/**/*.info' |
./node_modules/coveralls/bin/coveralls.js; fi

/home/travis/build/HabitRPG/habitica/node_modules/coveralls/bin/coveralls.js:18
        throw err;
        ^
Bad response: 422 {"message":"Couldn't find a repository matching this job.","error":true}

The command "if [ $COVERAGE ]; then ./node_modules/.bin/lcov-result-merger
'coverage/**/*.info' | ./node_modules/coveralls/bin/coveralls.js; fi" exited with 1.
store build cache

nothing changed, not updating cache
Done. Your build exited with 1.
```

**Comment by paglias** [comment]
@Yutsuten that's a weird error but it seems unrelated from the actual tests, can you merge in the latest develop commit?

**Comment by paglias** [comment]
Now the tests are passing

File: website/server/models/group.js
Comment (paglias) [inline website/server/models/group.js]: can you cache the user.preferences.language value in a variable at the beginning of this function? it'll make the code more readable. Like `let lang = user.preferences.lang`

Comment (paglias) [reply website/server/models/group.js]: Also group.chat[i] at the beginning of the for

**Review by paglias — CHANGES_REQUESTED** [review]
Left two comments, this is going to need tests before it can be merged

**Comment by Yutsuten** [comment]
This needs automated tests? Is it possible to test translations?

**Comment by paglias** [comment]
I would like to have unit tests for `translateSystemMessages` and yes, you should be able to tests translations, you can set the user language to a different language and then check with the helper in test/helpers/translate.js that the correct string is returned

**Comment by Yutsuten** [comment]
Nice, but for now I only have the english version of most strings.
Or with it I am able to create strings in another languages dinamically?

**Comment by paglias** [comment]
Sorry, what we care about is that given a message with an info object we can generate a correct string, so testing with english is fine, the important is that no "Error processing string" error is returned

**Comment by Yutsuten** [comment]
Looks like an unrelated integration test failed on Travis.
Tested locally and it passes, so can someone restart it please?

**Comment by paglias** [comment]
Hi @Yutsuten I've been speaking with @TheHollidayInn about this PR and we've decided that given the work underway at #9703 that's going to change a lot the way the Chat works we're going to wait until that PR is done and merged before merging this one

**Comment by Yutsuten** [comment]
@paglias Got it, no problem!
By the way, could I ask you a favor? When that PR get merged in develop, could you leave a message here notifying me? I was seeing the changes there and it *will* cause conflicts, so I want to fix those as soon as possible. Thank you!

**Comment by paglias** [comment]
Will do!

**Comment by Yutsuten** [comment]
Oh, it was merged already.

I've just updated my local install (branch `develop`), it took more time to make it work than I would like to admit. `npm run client:dev` was failing consistently, but at the end I finally was able to fix it by `npm install`ing it after downgrading node from `8.11.1` to `8.10.0`.

The next step will be fixing the conflicts...

**Comment by Yutsuten** [comment]
I was checking the conflict, and I noticed that the system message for quest start has been removed. Should I remove from my PR too?
I don't really understand the reason for removing this message, so just confirming here...

**Comment by Yutsuten** [comment]
Oh sorry, my mistake. The code for sending this message wasn't removed, but moved. Please ignore my message.

File: website/server/models/group.js
Comment (paglias) [inline website/server/models/group.js]: This function is really difficult to read, I'd suggest a few changes:

- put `group.chat[i].info` inside a variable so it's easier to access without having to always type `group.chat[i].info `
- add blank lines before the `for` loop, after each `case` statement
- split the function into at least two one with the `for` loop and one to translate a single message

File: website/server/models/group.js
Comment (paglias) [inline website/server/models/group.js]: let' use `chat.forEach` instead of `for (`

File: website/server/models/group.js
Comment (paglias) [inline website/server/models/group.js]: the signature of this function should be changed: instead of having 4 paraneters and passing `null, null` a lot of times it should be changed to accept a single object parameter:

```
schema.methods.sendChat = function sendChat (options = {}) {
    const {message, user, metadata, info = {}} = options
    ....
}
```

**Review by paglias — DISMISSED** [review]
The code works but it's really complicate so I left some comments on how to simplify it

File: website/server/models/group.js
Comment (Yutsuten) [reply website/server/models/group.js]: OK: `group.chat[i].info` -> `info`
OK: blank lines before `for` and after `case`
Against: splitting the function.

Reason 1: I know the function is long, but I see no meaning into creating a new function for translating a single message. Translations itself already happen in a single line of code. 

Reason 2: Even if I create a new function for generating messages, the number of lines in `_translateSystemMessages` will keep the same. There is no reason to put `shared.i18n.t( ... )` into a new function.

```javascript
// Unnecessary step, should call shared.i18n.t directly!!
function translateMessage(msg, ...parameters) {
  return shared.i18n.t(msg, parameters);
}
```

Reason 3: The function already does a single thing. It (1) initialize variables, (2) has a `for` loop for translating, (3) returns the result.

Please tell me if you are seeing a solution I can't. I personally see no reason/way to split the function into two or more.

File: website/server/models/group.js
Comment (Yutsuten) [reply website/server/models/group.js]: OK

File: website/server/models/group.js
Comment (paglias) [reply website/server/models/group.js]: @Yutsuten sorry I think I didn't explain very well, I wasn't suggesting creating a new function for each message. But more something like this:

```
function _translateSystemMessages (group, user) {
    ...
    for or foreach ( {
        translateMessage() // new function
    }
}
```

this way we have a function that only does the translation part

Comment (Yutsuten) [reply website/server/models/group.js]: Oh I see, I misunderstood you. I'm sorry.
So I should put the `switch` part inside the new function, returning the message. Makes some sense. Thank you for clarifying, I'll do this!

File: website/server/models/group.js
Comment (Yutsuten) [reply website/server/models/group.js]: OK

Comment (Yutsuten) [reply website/server/models/group.js]: I was about to implement those suggestions today, but this change in particular breaks a lot of tests, and requires changing code in a lot of scripts.
Maybe I complete it by this weekend...

**Comment by Yutsuten** [comment]
I've implemented the changes @paglias .

But there are 2 tests I wasn't able to rewrite fully...
Those tests deals with random data (random items dropped and a random user name).
Now these tests are as follows:

lines 505-513 `test/api/v3/unit/models/group.test.js`
```javascript
expect(Group.prototype.sendChat).to.be.calledWithMatch({
  // message: "`Participating Member found \d* Broken Twigs, \d* Tracks.`",
  info: {
    // items: { branches: \d*, tracks: \d* },
    quest: 'evilsanta2',
    type: 'user_found_items',
    user: 'Participating Member',
  },
});
```

Lines 130-137 `test/api/v3/integration/quests/POST-groups_groupid_quests_abort.test.js`
```javascript
expect(Group.prototype.sendChat).to.be.calledWithMatch({
  // message: /aborted the party quest Wail of the Whale.`/,
  info: {
    quest: 'whale',
    type: 'quest_abort',
    // user: '1526651788146ce91216'
  },
});
```

Any suggestion of how could I accomplish to make those tests test everything?
Tried here for hours with no success (not so much experience with testing though)...

**Comment by paglias** [comment]
schermata 2018-05-18 alle 16 52 43
schermata 2018-05-18 alle 16 53 07

Are these the two tests?

They seem to be using regular expressions to match the string

**Comment by paglias** [comment]
I haven't tried but I think that you should be able to replicate the previous behavior with something like this:

```
    expect(Group.prototype.sendChat).to.be.calledWithMatch({
      message: sinon.match(/aborted the party quest Wail of the Whale.`/),
      info: {
        quest: 'whale',
        type: 'quest_abort',
        // user: '1526651788146ce91216'
      },
 });
```

for the abort test, for the other since there are two expectations

```
    expect(Group.prototype.sendChat).to.be.calledWithMatch({
      message: sinon.match(/`Participating Member found/).and(sinon.match(/\d* (Tracks|Broken Twigs)/)),
      info: {
       ....
      },
 });
```

**Comment by Yutsuten** [comment]
Tested here, worked. Thank you!
How about the commented lines inside the `info` object? Should I leave as is or remove?

**Comment by paglias** [comment]
@Yutsuten you mean the items and user keys? If the values remain the same across tests and you can add a reliable expectation add it, otherwise you can remove the comments

**Comment by Yutsuten** [comment]
3 weeks since last update, please don't forget about this PR.

**Comment by paglias** [comment]
@Yutsuten sorry about this but we're dealing with some perf issues so right now we're not merging any new feature but only bugfixes

**Comment by Yutsuten** [comment]
@paglias got it, thank you for notifying!

**Comment by Yutsuten** [comment]
This is taking a lot of time... @paglias there is any estimate of when those performance issues will be fixed? Within weeks? Or months?

**Comment by paglias** [comment]
@Yutsuten no, I'm sorry but there is not ETA. I think after the chat part of this issue https://github.com/HabitRPG/habitica/issues/10472 is done we can merge this, not before, and it could be months. Right now we can't accept new features like this that add complexity or could slow down the server as we have big performances issues, I'm sorry

**Comment by SabreCat** [comment]
@paglias Should we move this to `status: on hold` or close it, then? With apologies to @Yutsuten !

**Comment by paglias** [comment]
@SabreCat I think keeping it in review, maybe with an extra on hold tag, is the best way There's nothing wrong with the PR but as it's a sizeable feature I don't think it's a good idea to merge it given all the issues and performance problems we have. But if you want to merge it it's fine for me

File: website/server/models/group.js
Comment (Yutsuten) [inline website/server/models/group.js]: After almost 2 hours fixing conflicts, just noticed that the `user` sent to `Group.toJSONCleanChat` as parameter don't have the `language` information anymore. Without this information I can't do any translation.
May I do a query to the DB to take the user's language?

P.S.: More 4 months and will be my PR's birthday, 1 year old!! hahah

Comment (paglias) [reply website/server/models/group.js]: @Yutsuten I'm so sorry about this but we've not yet had the time to make the changes to the chat routes necessary for this PR. Don't worry about fixing the conflicts, when we have a chance and the other issues are fixed we'll fix them Sorry again

Comment (Yutsuten) [reply website/server/models/group.js]: Got it, thank you for the reply! I think I'll push the conflict fixes anyway, some of them were tricky to solve, and it may help you guys when merging. The only problem now is that the `lang` variable I mentioned receives `undefined`.

**Comment by SabreCat** [comment]
@paglias We circled back to this in a meeting today--what would you say the situation is on this? Is there some groundwork we still need to do?

**Comment by paglias** [comment]
@SabreCat groundwork no, it should be ready for testing/merging

**Comment by Yutsuten** [comment]
Not sure if already fixed in another PR, but `user.preferences.language` used within the translation algorithm was returning `undefined` as mentioned [here](#discussion_r221954535).

**Comment by paglias** [comment]
@SabreCat do you want to merge this then?

**Comment by Alys** [comment]
@Yutsuten Do you still need an answer to your question about `user.preferences.language`?

**Comment by Yutsuten** [comment]
@Alys My objective was to inform to the habitica team about a possible problem, so you are acknowledged about it. No need to answer me, just thought that it would be something to be considered when deciding about merging or not.

If it is decided that it should be fixed, and the one who will fix it is me, an answer would be appreciated though.

**Comment by Yutsuten** [comment]
@Alys @paglias @SabreCat @lemoness This PR is open for more than a year. It took a quite amount of effort to make it. I did everything I was asked for.
I would be really grateful if the Habitica team stop ignoring my PR.

**Comment by SabreCat** [comment]
Yeah, not sure why this vanished from our radars for so long! :o I'm sending it up to a testing server now, should be able to merge once our mods and volunteers have had a chance to give it a try.

File: website/server/models/group.js
Comment (paglias) [inline website/server/models/group.js:347]: why is this function static? Wouldn't make more senso to be a class method? so you can call group.translateSystemMessages(user) and the group will be available as `this`

File: website/server/models/group.js
Comment (paglias) [reply website/server/models/group.js]: Is this still an issue? Can you point to me where it's happening?

**Comment by paglias** [comment]
The issue with the test comes from the fact that in `_translateMessage` the passed info object doesn't contain the user

**Comment by paglias** [comment]
Schermata 2019-05-10 alle 15 53 50

It's called from here.

In the case of the failing spell issues `info` only contains `{ type: 'spell_cast_party' }` so this code is not working `spells[info.class][info.spell].text(lang)`

I was not able to replicate the issue where the `lang` was undefined

Also, can we move all these new functions to a new file under server/libs for example? it's difficult to read it now given how big the group modal has become

File: website/server/models/group.js
Comment (paglias) [inline website/server/models/group.js]: this should be moved after the toJSON conversion or it'll modify the mongoose object directly which we don't want to do

**Review by paglias — CHANGES_REQUESTED** [review]
Found some issues (see the comments as well)

**Comment by Yutsuten** [comment]
@paglias Thank you for checking the code. Looks like a lot of problems appeared while this PR was on hold...

I'll try to do something about those problems. I haven't changed the Habitica's source code for a while, so it may take some time to be able to fix them.

**Comment by paglias** [comment]
@Yutsuten no worries, it was our fault, let me know if you need any help or have any question

File: website/server/models/group.js
Comment (Yutsuten) [reply website/server/models/group.js]: It was a problem on Oct/2018, but don't know how it is now. I'll have to build Habitica locally again to check. If it is still an issue, I'll write here.

**Comment by Yutsuten** [comment]
This time I am going to setup my local Habitica with Docker. At work, I use `docker` and `docker-compose` at a daily basis. I have already created some personal projects with them too. So I think it is ok to say that I have some experience on it.

BUT, please pardon me but it may sound rude.

The current setup with docker on Habitica sucks. `git clone` inside `Dockerfile` is unacceptable. `Dockerfile-Dev` and `docker-compose.dev.yml` seems to be a work in progress, but still unusable for development. In my opinion, _everything_ related to docker needs to be rewritten from scratch.

I'll do it.

If you have interest in the new docker setup, please tell me and I open a new Pull Request with the new `Dockerfile` and `docker-compose.yml` files. The [wiki about docker](https://habitica.fandom.com/wiki/Setting_up_Habitica_Locally#Docker) will need a rewrite too, I may do it if it sounds good. It will be as easy as typing `docker-compose up -d` on the bash (no more commands needed), so probably it will become the easiest and smallest guide at [Contributing Setting up Habitica Locally](https://habitica.fandom.com/wiki/Setting_up_Habitica_Locally).

**Comment by paglias** [comment]
Regarding the Dockerfile we'd be happy if you can improve the setup! We know it's not up to date but we don't use Docker for development but only for production use (the `Dockerfile` file). So that should not be changed because it's required for our own production deployments.

But the other files can be modified with no issue

File: website/server/models/group.js
Comment (Yutsuten) [reply website/server/models/group.js]: Updating the `group` after the `let toJSON` isn't meaningless?
The `group` variable isn't used anymore after `let toJSON = group.toJSON();`.

File: website/server/models/group.js
Comment (Yutsuten) [reply website/server/models/group.js]: Just confirmed that the problem _still_ exists with the latest commit of this PR.

After editing the mentioned part of the source code to
```javascript
function _translateSystemMessages (group, user) {
  console.log('####################');
  console.log(user);
  let lang = user.preferences.language;
  console.log(lang);
  console.log('####################');
```

I accessed the party page and got this output:

```
server_1  | ####################
server_1  | { contributor: {},
server_1  |   party:
server_1  |    { quest:
server_1  |       { progress: [Object],
server_1  |         RSVPNeeded: false,
server_1  |         key: 'basilist',
server_1  |         completed: null },
server_1  |      order: 'level',
server_1  |      orderAscending: 'ascending',
server_1  |      _id: '066ceec3-9eb8-45e7-852b-a07a9202f6a8' },
server_1  |   guilds: [],
server_1  |   _id: 'ef191782-0b23-4240-a2b4-aacb0ba0d99d' }
server_1  | undefined
server_1  | ####################
```

Any recommended way to access the user's language in this case?

File: website/server/models/group.js
Comment (Yutsuten) [reply website/server/models/group.js:347]: Changed to a class method here: 0360326f41485c8a9514b586dc4d9e4ef30d587c
It will break some tests, but I agree with you that it is better this way.

Now `user.preferences.language` raises an exception for me.

```
server_1  | 2019-05-15T06:37:29.319Z - error: TypeError: Cannot read property 'language' of undefined
server_1  |     at model.translateSystemMessages (/code/website/server/models/group.js:532:31)
server_1  |     at Function.groupToJSONCleanChat [as toJSONCleanChat] (/code/website/server/models/group.js:422:9)
server_1  |     at process._tickCallback (internal/process/next_tick.js:68:7)
```

To avoid the exception, for now I'm doing a verification if `user.preferences` is not null.

```javascript
let lang = user.preferences ? user.preferences.language : 'en';
```

Comment (paglias) [reply website/server/models/group.js:347]: The issue is that for some reason the user's preferences aren't loaded from the db because they always exist so it looks likes there are cases where user.preferences aren't in the list of user fields loaded which is not fine if we need to access the property.

It's probably the spell casting route that doesn't load them

File: website/server/models/group.js
Comment (paglias) [reply website/server/models/group.js]: If you modify the group object before serializing it to json it means that you're changing the mongoose class representing the group (so the in memory representation of the db data) which we don't want to do since translated messages should never end up in the db.

Instead we want to first convert it to json `let toJSON = group.toJSON(); ` and then translate that object with `_translateSystemMessages(toJSON, user); `

Comment (Yutsuten) [reply website/server/models/group.js]: But after serializing it into JSON, it is not saved to the database. For testing, I changed the language here and the database didn't change at all.
All is done just to show the messages to the user, there is no save involved.

And if there is a case where it would be saved, why did you ask me to make the `translateSystemMessages` a class method?
The `group` object wouldn't be needed anymore...

By the way, I already changed the method to a class method.

Comment (Yutsuten) [reply website/server/models/group.js]: One more thing, If there is such possibility of `group` being saved, I think the call to `getGroupChat` is a bigger problem than my call to translate the messages.

File: website/server/models/group.js
Comment (paglias) [inline website/server/models/group.js:360]: I've lost the previous conversation but going back to this: I know that the group object is not going to be saved after this but it's bad practice to apply changes that will never end up in the db to a db document class and across the entire project we first serialize to a plain js object and then change what's needed

**Review by paglias — DISMISSED** [review]

File: website/server/models/group.js
Comment (Yutsuten) [reply website/server/models/group.js:360]: So the class method is not needed anymore... Some wasted work and time detected.

Comment (paglias) [reply website/server/models/group.js:360]: Well you can still use it and have it return the serialized data

Comment (Yutsuten) [reply website/server/models/group.js:360]: Changed back to static method, and now it changes the serialized object. ab0dae8df359ae64d083b97e54d1e9bb4f93e95d

Hope you're done with this kind of reviews, I think I've already worked waaay too much on this, more than I should.

If there are any other improvements you find, please do it after merging this and in a new PR.

Comment (paglias) [reply website/server/models/group.js:360]: > Hope you're done with this kind of reviews, I think I've already worked waaay too much in this, more than I should.

As I said previously we're sorry it took so long with this PR and if you don't want to work on it anymore that's fine and someone from the staff will finish it but I'm not going to merge it if there are still issues, I'll give it another look later but at least there's the problem with the preferences object being undefined in some case

Comment (Yutsuten) [reply website/server/models/group.js:360]: > someone from the staff will finish it

The preferences problem is the only thing missing. If I change the `lang` variable manually, I can check that everything works as expected. There is no missing feature here.

And it is not about me working on it or not. I only want this PR to be merged. That's why I created it, after all.

When I said "this kind of reviews", I meant refactorings you ask. "_It should be implemented this way and not that way_" or "_It is better this way_". I really think that this is not the kind of thing you should write in a 1 year 3 months old PR where the owner changed EVERYTHING he was asked to do so far, and always within a week.

**Lets focus on creating a working feature.** There is no need to merge it only then it becomes 100% _perfect_. Improvements and refactorings can be done later, by the Habitica team, as you said. But on another PR, please.

**Comment by paglias** [comment]
I've investigated the issue locally, turns out the language was never loaded when getting a group because `preferences` wasn't part of the list of fields to load when getting a group. 

This means that if we deployed this everyone would have seen the messages in English since the language wasn't available when loading the group and chat on the website (on mobile a different route, not affected by this issue, was used).

I fixed the problem (and did some refactoring) in https://github.com/HabitRPG/habitica/pull/11170 which will be tested for a few days and then merged if everything looks good.

Of course we'll give you the credits for the feature.

Regarding you last comment at https://github.com/HabitRPG/habitica/pull/10019#discussion_r284242230

> Lets focus on creating a working feature. There is no need to merge it only then it becomes 100% perfect. Improvements and refactorings can be done later, by the Habitica team, as you said. But on another PR, please.

As shown the feature wasn't really working. Regarding the refactoring I try not to be too picky about PRs from contributors because we know no one is getting paid for them and we're very grateful for them. But in this case we're talking about a feature that adds almost 500 lines of new code to an already complex system so I was trying not to increase our tech debt.

I'll post here once it's been tested, merged and the notated in your Habitica profile. Thanks

**Comment by Yutsuten** [comment]
> But in this case we're talking about a feature that adds almost 500 lines of new code to an already complex system so I was trying not to increase our tech debt.

You've got a point here.

> As shown the feature wasn't really working.

Getting the language was missing, I know. The ONLY thing missing, from my point of view.

---

Anyway, I saw you created a new PR and did some work there. Good job.
Hope you're not mad at me. I know I wrote more than I should.

Just hope this feature gets released soon...

**Comment by paglias** [comment]
No worries, we're going to give it some testing for 1-2 weeks and then release it, I'll ping you when it's done

**Comment by Yutsuten** [comment]
Nice to hear that, thank you!

**Pull request merged**
